Block Editor: Allow control over drop cap feature with `useEditorFeature` helper

gziolo opened the pull request:
## Description

Related to https://github.com/WordPress/gutenberg/issues/21583.

So far there is one new field proposed in the editor (and block editor) settings on the client: `__experimentalFeatures`. It's exposed from the server.

Usage:

```js
// edit.js

const Edit = ( props ) => {
    const isDisabled = ! useEditorFeature( 'typography.dropCap' );
    // ...
};
```

~The discussion about the shape of config files needs to happen first before we commit to the more solid proposal.~
There was a very in-depth discussion about the shape of the theme config (and the WordPress core config with defaults for themes) in this thread: https://github.com/WordPress/gutenberg/pull/22291#pullrequestreview-411615000. We didn't commit to the final shape but we have enough to continue work on this feature and similar concepts related to themes and way to have more control over the block editor.

Finally, I tried to move `supports` from the JS to `block.json` for all blocks in https://github.com/WordPress/gutenberg/pull/22422. At the same time, I wanted to try to rename it to `features` but it ended up as a very deep refactoring so I took step back and figure out why not put `features` as a section of `supports` that has quite good support in the codebase. Based on that idea I included block overrides for features using the following definition of the block:

```js
registerBlockType( 'my/block', {
    supports: {
        __experimentalFeatures: {
            typography: {
                dropCap: true,
            }
        },
    },
    // ... everything else
} );
```

## Open questions

There is still this remaining question whether we need to define default settings in WordPress core that can be overridden by themes on the server using JSON format. An alternative approach would be to use PHP array or move that to the client in the `getSettings` implementation.

## How has this been tested?

First, I ensured that drop cap functionality still works for the Paragraph block as before.

I also manually modified `dropCap` flag in Paragraph block by modifying the `supports` field to ensure it disables the feature when it is set to `false`. I also removed the section for Paragraph block and tested that global settings toggle the features properly.

## Types of changes
New feature (non-breaking change which adds functionality).

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [N/A] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [N/A] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [N/A] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

gziolo committed:
Block Editor: Integrate theme config with useEditorFeature

nosolosw commented on lib/theme.php:
```diff
@@ -0,0 +1,55 @@
+ * @return array Filtered editor settings.
+ */
+function gutenberg_extend_settings_features( $settings ) {
+	$theme_features = gutenberg_experimental_get_theme_config(
```
Can we find a way to pass `gutenberg_experimental_get_theme_config` all the keys that we want to retrieve so that we only execute this function once? Reading from the file system and `json_decode` can be expensive operations.

gziolo replied:
How about we don't use JSON at all for the default theme config? 😄 

I'm serious here. If you combine this feedback with https://github.com/WordPress/gutenberg/pull/22291/files#r423893362 – the issue with the JSON file in the root folder, it raises the question of why we opted for JSON for WordPress core in the first place. Do we plan to share this file with the client?

nosolosw replied:
To be honest, I don't remember. Perhaps in some early explorations we pass it directly to the client? It seems fine to me that we go with PHP for this. cc @jorgefilipecosta @ItsJonQ in case they have better memories.

nosolosw commented on packages/editor/src/components/provider/index.js:
```diff
@@ -113,12 +113,14 @@ class EditorProvider extends Component {
 				'__experimentalEnableFullSiteEditing',
 				'__experimentalEnableFullSiteEditingDemo',
+				'__experimentalFeaturesConfig',
+				'__experimentalEnableFullSiteEditingDemo',
```
```suggestion
```

nosolosw commented on lib/theme.php:
```diff
@@ -0,0 +1,55 @@
+	$settings[ '__experimentalFeaturesConfig' ] = $theme_features;
+
+	$theme_blocks = gutenberg_experimental_get_theme_config(
+		dirname( dirname( __FILE__ ) ) . '/experimental-default-theme.json',
```
As per https://github.com/WordPress/gutenberg/pull/19883#discussion_r392226550 we need to move this file somewhere under the `lib/` folder, otherwise, it won't be included in [plugin ZIP](https://github.com/WordPress/gutenberg/blob/master/bin/build-plugin-zip.sh#L114). I had a few PRs that also bundled this change but didn't land yet, so if we are going to use it for the block-editor settings, can we perhaps do it here already?

gziolo replied:
See https://github.com/WordPress/gutenberg/pull/22291#discussion_r424259262.

gziolo replied:
Moved to `lib`.

nosolosw commented on experimental-default-theme.json:
```diff
@@ -0,0 +1,23 @@
+			"dropCapEnabled": false
+		}
+	},
+	"blocks": {
```
I'm thinking that we probably want to group this theme.json _per feature_ to ease extraction (styles, features, etc) and each key would have a "global" and a "blocks" subkey.

Example:

```
{
    "styles": {
        "global": {
            "color": { ...}
        },
        "blocks": {
            "core/paragraph": {
                "color": { ... }
            }
        }
    },
    "features": { // or config, editor, or something like that
        "global": { ... },
        "blocks": {
            "core/paragraph": { ... }
        }
    }
}
```

For context, [the shape](https://github.com/WordPress/gutenberg/issues/21583#issuecomment-613911005) Riad originally proposed, to which I [suggested](https://github.com/WordPress/gutenberg/pull/22076#issuecomment-624183378) using "styles" for the things relative to the global styles.

nosolosw replied:
I think I'd like even more something with less hierarchy: 

```
{
    "styles": {
        "global": {
            "color": { ...}
        },
        "core/paragraph": {
            "color": { ... }
        }
    },
    "features": { // or config, editor, or something like that
        "global": { ... },
        "core/paragraph": { ... }
    }
}
```

This way, when we extract the data in `useEditorFeature`, it's as easy as:

```js
has( getSettings(), [
    '__experimentalConfig', // we only need one key for this
    context, // 'global' or the block name
    feature,
] )
```

gziolo replied:
@nosolosw, great idea to group it like this when passing to the block editor as part of settings object. How would you structure the `theme.json` to optimize for the person that needs to maintain it? 😄 I think those two not necessarily need to align.

gziolo replied:
@nosolosw, I think you answered in https://github.com/WordPress/gutenberg/pull/22291#discussion_r425107659 :) Le's fold this discussion and continue there.

mtias commented on experimental-default-theme.json:
```diff
@@ -0,0 +1,23 @@
+	},
+	"features": {
+		"typography": {
+			"dropCapEnabled": false
```
I think we could drop the "Enabled"

youknowriad commented on lib/experimental-default-theme.json:
```diff
@@ -0,0 +1,23 @@
+			"dropCap": false
+		}
+	},
+	"blocks": {
```
The question here is whether "features"/"config" and "style" should have inner keys "global", "blocks" or the opposite. I don't  feel strongly but  I feel we should be consistent between "style" and "features"

gziolo replied:
From the perspective of the theme maintainer, it might be easier to group by block because you can have tons of them and their names can be long like or hard to remember. Words like "features" or "styles" are short and easy to include. However, I don't have strong preferences. Whatever works for the majority is fine here. We can change it also later and map it in PHP or JavaScript. It should be all transparent from the perspective of `useEditorFeature` consumer.

nosolosw replied:
I've just finished the last iteration of https://github.com/WordPress/gutenberg/pull/20290 and went with `styles` as the top-level, but will rework it after lunch to use the block names (+ `globals`) as the top-level keys.

nosolosw replied:
oh, before I go down the path, I was thinking we'd have this:

```
{
    "global": {
        "styles": { /* not sure what can go here, nothing or perhaps things we attach to the root container like background color ?*/ },
        "preset": { /* default presets go here */  },
        "features": { /* things enabled by this PR go here */ }
    },
    "core/paragraph": {
        "styles": { /* block styles */ },
        "preset": { /* presets per block, if this is something the block can do */  },
        "features": { /* things enabled per block */ }
    },
    "core/group": {
        "styles": { /* block styles */ },
        "preset": { /* presets per block, if this is something the block can do */  },
        "features": { /* things enabled per block */ }
    }
}
```

Re-reading Riad's message above, you seem to suggest that we can group all blocks under a "blocks" key, right? I think I have a slight preference to avoid that if that's ok.

gziolo replied:
@nosolows, your example is exactly what I was coming to lately 😃

youknowriad replied:
That example works for me. I'm not sure whether "features" or "presets" are enough for all "configs" that's why I'm hesitant between having these or a single "config" holding everything.

Put differently, what config will we have:

 - Enable/disable things
 - Provide presets
 - Anything else?

youknowriad replied:
Also how do you define the "Global  styles" UI in this config.

gziolo replied:
@nosolosw, can you expand `global` or one of the blocks with a more detailed example? 

Just to make sure we are on the same page, this proposal is targeting `theme.json` that every theme will be able to define.

In the WordPress core, we probably need only the `global` section to provide default values for everything. 

What's the story for core blocks and 3rd party blocks, where would they define `features` or `styles? Does it make sense to allow them in `register_block_type` as part of `settings` object? Can we automatically read them from `theme.json` and inject overrides into existing block definitions on the server and on the client?

nosolosw replied:
I guess it'd look like something like this:

```
{
    "global": {
        "features": {
            "dropCap": false,
        },
        "presets": {
            "color": [ /* defines color presets globally */ ],
            "font-size": false, /* disabled globally */
        },
    },
    "core/paragraph": {
        "styles": {
            "color": "red",
            "font-size": 13
        },
        "presets": {
            "font-size": [ /* define font size presets, so they are enabled for this block */ ]
        },
        "features": {
            "dropCap": true
        }
    },
    "core/navigation": {
        "presets": {
            "color": [ /* defines a different color palette */ ]
        }
    }
}
```

It absorbs the things [mentioned here](https://github.com/WordPress/gutenberg/issues/21583#issuecomment-613911005), although I'm unsure about the "global styles ui" bit that @youknowriad mentions. My thinking is: that UI is not different from the block-level inspector controls, part of it should be generated from the implicit block atts of the blocks (line-height controls, etc), the rest is static. Does that sound about right?

However, that makes me think: would we need different configs for different editors (post, site, widgets)?

youknowriad replied:
In your example you're disabling the font-size in the "presets". Shouldn't this be done in "features"? Maybe a hint that presets and features are the same thing "config"

youknowriad replied:
> My thinking is: that UI is not different from the block-level inspector controls, part of it should be generated from the implicit block atts (line-height controls, etc), the rest is static. Does that sound about right?

In terms of UI, I originally was thinking that UI  should be generated but I changed my mind :) and now more leaning towards the theme saying what things should be editable by UI (palette, a random CSS variable representing a color and used in theme.json in different places..., a random font size CSS variable).

nosolosw replied:
Yeah, I did it intentionally to have this kind of discussion. I know that at the current API is different for disable them & configure its values, but I thought that if we use a config file to do this we could benefit from some compression (a single place to do everything).

The only reason I like `presets` at the secondary level is that the data they hold can be long (array of many things with names, slug, etc), unlike the rest of features where most of them would be booleans.

youknowriad replied:
At this point of the discussion, I have all my concerned shared and discussed. I'm happy if we move forward with a format and implement things. We didn't decide to remove the "experimental" yet :)

nosolosw replied:
> We didn't decide to remove the "experimental" yet :)

I should take that win! 🙈 :tada: 

I feel more comfortable about the direction and format now, we could open it in a release or two. The only bit I'd like to confirm before removing the `experimental` would be whether we want to keep JSON or use PHP as Grzegorz pointed out.

youknowriad replied:
too many decision to make :) 

I don't want to make it "non-experimental" too soon for other reasons too. We could also decide to make some changes  to block markup for themes using theme.json (button markup simplification, alignments, color palette hook removal...)

nosolosw replied:
> the theme saying what things should be editable by (global styles) UI

So I'm thinking that the whole `theme.json` is tailored to control "the editor", but we haven't defined what's "the editor". I'm not sure we need this, but what you said above sounded to me like we wanted different things for the post editor and the site editor (where global style is). Not sure we should take any decision now, we can perhaps start with what we have and decide later. Thought about raising it anyway.

youknowriad replied:
for me theme.json  is about the theme (so the whole  website) and not just the "editor'.

jorgefilipecosta replied:
I'm not sure if we should have a "global" property. To me, everything we set on theme JSON is global unless we nest it inside a block/template part/slector.

gziolo replied:
@jorgefilipecosta, are you proposing something like:

```
{
    "styles": { /* not sure what can go here, nothing or perhaps things we attach to the root container like background color ?*/ },
    "presets": { /* default presets go here */  },
    "features": { /* things enabled by this PR go here */ }
    "overrides": {
        "core/paragraph": {
            "styles": { /* block styles */ },
            "presets": { /* presets per block, if this is something the block can do */  },
            "features": { /* things enabled per block */ }
        },
        "core/group": {
            "styles": { /* block styles */ },
            "presets": { /* presets per block, if this is something the block can do */  },
            "features": { /* things enabled per block */ }
        }
    }
}
```

This format makes it look more like config for Babel or ESLint :)

nosolosw replied:
I actually like more the proposal with global, the config looks simpler both from a human and programmatic perspective (see, [it only requires a loop](https://github.com/WordPress/gutenberg/pull/20290/files?file-filters%5B%5D=.php#diff-4339a9e86722fd029c231f3fb65f0ad2R287)).

jorgefilipecosta replied:
I'm proposing a shape similar to this one:
```
{
		"styles": {
			"color": "blue",
			"font-size": "14"
		},
		"presets": {
			color:
			gradient:
		},
		"core/paragraph": {
			"styles": {
				"color": "blue",
				"font-size": "14"
			}
		}
}
```

More details at https://github.com/WordPress/gutenberg/pull/20290#pullrequestreview-413787582.

But in summary, the shape of global styles should be equivalent to the shape we have inside a "core/group" or "core/template-part" selectors. If inside the core group/"template part" we don't have a global property the global shape should also not have them.

gziolo replied:
@jorgefilipecosta thanks for sharing. I see one challenge with your proposal. You need to keep the list of all keywords that aren't blocks or template parts to be able to divide what's in the root of the config file.

nosolosw replied:
Ah, I see the logic to that proposal, thanks for sharing, Jorge.

My main concern with that proposal would be that I'd like to explore different ways of declaring nested content. I am not too keen on using that approach as it has scale issues with nested levels. I've tried it and it becomes cumbersome to declare only 2 nested levels, 3 are quite unwieldy. In my past experiments with a simple pattern like the "numbered features" I needed 3 levels to be able to encapsulate the pattern, and I presume that's going to be common in some situations if people use the columns block. However! What I learned the past week is that nested content is a huge topic and a discussion for later, so I don't want to make many assumptions just yet.

Note also a big disclaimer here: we can iterate on the shape of theme.json. I'm eager to start with something so we can gather feedback from theme-authors asap on the parts that are missing.

jorgefilipecosta replied:
> You need to keep the list of all keywords that aren't blocks or template parts to be able to divide what's in the root of the config file.

We already have the list it is "styles" and "presets" everything else is blocks. Keeping a list is not hard as if we added another thing different from styles or presets we need to process the thing so we can easily add to the list.

If we really want to avoid the list another option would be to make the "blocks" inside a blocks key: 
```
{
		"styles": {
			"color": "blue",
			"font-size": "14"
		},
		"presets": {
			color:
			gradient:
		},
		"blocks": {
			"core/paragraph": {
				"styles": {
					"color": "blue",
					"font-size": "14"
				}
			}
		}
}
```

I would prefer a "blocks" key than a global key, as it would keep the shape on nesting.

nosolosw replied:
Jorge and I talked about the shape of `theme.json`, and we've agreed to [land what we have](https://github.com/WordPress/gutenberg/pull/20290) (top-level blocks + global) and iterate. How we capture nested content is going to influence the shape of this, but the fact that it's experimental give us leeway to modify it without concerns about backward-compatibility. Probably it's a good idea to not remove the experimental until we figure out that part.

gziolo replied:
Fantastic, thank you for letting me know. I will update the default config to follow that (I still need to rebase). It looks like only need to move `features` to `global` section and leave the rest as is 👍

gziolo replied:
Now that we have https://github.com/WordPress/gutenberg/pull/22518, I updated this PR to follow the recommendations.

jorgefilipecosta commented on packages/block-library/src/paragraph/edit.js:
```diff
@@ -57,7 +57,7 @@ function ParagraphRTLToolbar( { direction, setDirection } ) {
 
 function useDropCap( isDropCap, fontSize, styleFontSize ) {
-	const isDisabled = useEditorFeature( '__experimentalDisableDropCap' );
+	const isDisabled = ! useEditorFeature( 'typography.dropCap' );
```
Would it make sense to use [ 'typography', 'dropCap' ] directly like we force lodash paths to be using lint rules?

gziolo replied:
I would prefer to avoid optimizing the hook based on the internal implementation that can change over time.

jorgefilipecosta commented on packages/block-editor/src/components/use-editor-feature/index.js:
```diff
@@ -1,23 +1,61 @@
 			const { getSettings } = select( 'core/block-editor' );
 
-			return getSettings()[ featureName ];
+			const path = featurePath.split( '.' );
```
I made a related comment if we should pass the array paths directly https://github.com/WordPress/gutenberg/pull/22291#discussion_r425313407.
But I guess an option would be to just pass the path to lodash functions this support 'a.b' format directly and we don't need to do the split. So we would support the array format and 'a.b' internally we should probably use arrays because of the performance reason.

gziolo replied:
```js
_.get(object, 'a[0].b.c');
// => 3
 
_.get(object, ['a', '0', 'b', 'c']);
// => 3

_.get(object, ['a', '0', 'b.c']);
// => undefined
```

You can't mix two approaches, that's why `split` is used. We can `join` those strings as an alternative and pass as a string to lodash.

jorgefilipecosta reviewed (commented):
I think the idea is that typography.dropCapEnabled may be enabled globally but, for example, disable inside a specific template part.
I'm not sure if we should include __experimentalBlocksConfig and __experimentalBlocksFeatures as settings. I imagine we may have a useGlobalStyles hook that queries the endpoint we will use to save and retrieve the global style.
That hook would be contextual, e.g., it would be aware that a paragraph inside a specific template part may have typography.dropCapEnabled disable while a paragraph not nested may have it enabled.
How to contextualize the global styles is not yet decided, so I guess provided we mark reverting as experimental for now, we can keep the current approach.

gziolo commented:
> I think the idea is that typography.dropCapEnabled may be enabled globally but, for example, disable inside a specific template part.

We want to have a way to give users control for this feature globally and on the block level for start. It's enough to handle. The idea is to offer `useEditorFeature` hook as the way to resolve it in the block editor so we could apply any type of filtering in the future. This PR illustrates how we could override the global setting for an individual block.

> I'm not sure if we should include __experimentalBlocksConfig and __experimentalBlocksFeatures as settings. I imagine we may have a useGlobalStyles hook that queries the endpoint we will use to save and retrieve the global style.

For this PR, it probably makes sense to remove `__experimentalBlocksConfig` which is something that is still discussed in relation to `theme.json` defined by the theme rather than Gutenberg itself. I believe that Gutenberg should only provide defaults that are included in `__experimentalFeaturesConfig` and we might even want to keep those defaults on the client rather than in another `theme.json` file inside the `lib` directory.

gziolo commented:
I changed slightly the goal of this PR to:
"Allow control over drop cap feature with `useEditorFeature` helper"

I tried to move `supports` from the JS to `block.json` for all blocks in https://github.com/WordPress/gutenberg/pull/22422. 

I hope it will make it easier to move forward with the proposal.

I removed the section with overrides from the default config. I included mostly for exploration and I think it should be part of the actual theme config, not the one with the defaults that themes can override and we should work on it separately.

There is still this remaining question whether we need to define default settings in WordPress core that can be overridden by themes on the server using JSON format. An alternative approach would be to use PHP array or move that to the client in the getSettings implementation.

youknowriad commented:
I'm not sure `__experimentalFeatures` key is needed in the example above as we already have top-level things. (align, __experimentalColor but happy to move forward and adapt as we add things.

gziolo commented:
> I'm not sure `__experimentalFeatures` key is needed in the example above as we already have top-level things. (align, __experimentalColor but happy to move forward and adapt as we add things.

It's my way to make the initial implementation more limited but I'm happy to flatten that if we prefer to keep everything at the top level.

aduth commented on lib/theme.php:
```diff
@@ -0,0 +1,54 @@
+		return $empty_config;
+	}
+	$theme_config = json_decode(
+		file_get_contents( $config_path ),
```
I'm getting the impression from the rest of this function that this is intending to be very precise in its validation of the configuration file being usable. One thing that's missed is that `file_get_contents` [will generate a warning](https://www.php.net/manual/en/function.file-get-contents.php#refsect1-function.file-get-contents-errors) if the file doesn't exist. Would we want to [absorb](https://www.php.net/manual/en/language.operators.errorcontrol.php) the warning? Or let the warning be logged?

gziolo replied:
> One thing that's missed is that file_get_contents will generate a warning if the file doesn't exist.

There is a check for the file's existence one line earlier.

> Would we want to absorb the warning? Or let the warning be logged?

In extended the existing logic a bit and I didn't think about warnings if the file couldn't be processed. It feels like the best idea would be to come up with a proper JSON schema validation for the file at some point. It will not only give proper feedback to the theme's consumer but it will also support other use cases like the ability to mark translatable fields for externals tools like WP-CLI (related discussion start here: https://github.com/wp-cli/i18n-command/pull/210#issuecomment-628545550.

aduth replied:
> > One thing that's missed is that file_get_contents will generate a warning if the file doesn't exist.
> 
> There is a check for the file's existence one line earlier.

Ah. In truth, I did miss it. There are apparently a couple other cases though:

>An E_WARNING level error is generated if filename cannot be found, maxlength is less than zero, or if seeking to the specified offset in the stream fails.

https://www.php.net/manual/en/function.file-get-contents.php#refsect1-function.file-get-contents-errors

gziolo replied:
True, there is always something that might go wrong. If this is how WordPress core handles it, I'll update code tomorrow 👍

gziolo replied:
I simplified handling and added `@` to silence potential warnings.

aduth replied:
Noting that late today, I stumbled by accident to this conversation:

https://make.wordpress.org/core/2020/03/20/updating-the-coding-standards-for-modern-php/#comment-38241

Which points to this guideline:

https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#error-control-operator

Even given the guidelines here, I think the usage should be appropriate, given that it's being acknowledged and the error case is handled by the subsequent condition.

aduth commented on packages/block-editor/src/components/use-editor-feature/index.js:
```diff
@@ -1,23 +1,51 @@
 			const { getSettings } = select( 'core/block-editor' );
 
-			return getSettings()[ featureName ];
+			return get( getSettings(), path, defaultValue );
```
I forgot that we removed the custom ESLint rule we had 😢 But it wouldn't have caught the trickery here anyways.

https://github.com/WordPress/gutenberg/pull/15139#discussion_r278505058

gziolo replied:
I know, but the alternative is to split the string and I'm sure it is comparable to what `get` does internally.

aduth replied:
> I'm sure it is comparable to what `get` does internally.

As with much of Lodash, you'd be surprised to find out just all it does 😄 

https://github.com/lodash/lodash/blob/master/.internal/stringToPath.js

I'm not actually suggesting any change here, though I do note the inconsistency here if it's something we think should be an ESLint rule in general.

Generally, I think it speaks more to how we've chosen to reference these properties in dot notation.

gziolo replied:
Yes, it was raised by @jorgefilipecosta in https://github.com/WordPress/gutenberg/pull/22291#discussion_r425313407. I feel resistance for using an array as a param in React hooks as it would be used as 2nd param in `useSelect` that is now:
```js
[ blockName, path ]
```

Now I think about it, we could cheat a bit and do:
```js
[ blockName, ...path ]
```

if we opt for a different API for `useEditorFeature`. You both can decide :)

Some alternative options:

1. With a problematic handling for the default value:
```js
const Edit = ( props ) => {
    const isDisabled = ! useEditorFeature( 'typography', 'dropCap', false );
    // ...
};
```

2. With an array as 1st param:
```js
const Edit = ( props ) => {
    const isDisabled = ! useEditorFeature( [ 'typography', 'dropCap' ], false );
    // ...
};
```

aduth replied:
I mean, there's a certain convenience of the dot notation, I'd grant that. Since it's something of a language hack, I'm not sure just how obvious the syntax is for most people. The array form isn't much better, though 🤷 

It could be just like Lodash where both are technically supported, and the string-form dot notation is casted into the array path. For what it's worth, I implemented a basic [`getPath` implementation](https://github.com/aduth/hpq/blob/master/src/get-path.js) for `hpq` that could be used. _(Though, as I look at it now, it should probably check truthiness of `object` before [operating via `in`](https://github.com/aduth/hpq/blob/c9704368e06be8cfc3d71696dee44d93f7e9dd87/src/get-path.js#L14), to avoid error)_

As you note, `path` could be spread into the dependencies. I have a sense we do this already in some places? I don't really see it as much of a cheat.

gziolo replied:
I want to move on with the proposal as it's blocking other efforts.

The more I think about it, the simplest approach would be to flatten the object structure to avoid the issue in the first place:

```js
const Edit = ( props ) => {
    const isDisabled = ! useEditorFeature( 'dropCap' );
    // ...
};
```
🙂 

Interestingly, it's something that was raised by @youknowriad or @nosolosw in different places but I'm not 100% sure if they were thinking about exactly that.

ZebulanStanphill replied:
Would it be possible for plugins to add "editor features"? If so, I would think we should probably use namespacing here, e.g. `'wp', 'drop-cap'` or something like that.

gziolo replied:
@ZebulanStanphill since it's using `supports` from the block definition, it's extensible by design. We can discuss it further as the work progresses. It's marked as experimental so it's all open for changes. I didn't think about namespacing so far mostly because there are so many existing features already considered stable so we would need to find migration path.

gziolo commented:
We discussed with @youknowriad that it would be nice to document the priorities:
> `theme.json`, `block.json`... which overrides which... and how to use these

I plan to follow @nosolosw and his work in #22518 and include a section to:

https://github.com/WordPress/gutenberg/blob/38155149ad7cf583d08044fd7654ce5fb1589d1e/docs/designers-developers/developers/themes/theme-json.md

gziolo commented:
I rebased this branch with `master` and applied all necessary changes. I also renamed `lib/theme.php` to `lib/editor-features.php`. Everything should operate as before.

gziolo commented:
> I still want to extend the new documentation page that touches `theme.json`, but it shouldn't be a blocker for this PR.

I opened #22622 that is focused on documenting all the changes proposed in this PR.

nosolosw commented on packages/block-editor/src/components/use-editor-feature/index.js:
````diff
@@ -1,23 +1,51 @@
+ * ```
  */
-export default function useEditorFeature( featureName ) {
+export default function useEditorFeature( featurePath, defaultValue ) {
````
I'm thinking that this component merges data from a few sources: block settings, editor settings, block.json, theme.json, filters, etc. Do we need to take a default value as well?

gziolo replied:
There are filters, 3rd party plugins, 3rd party blocks, you can't guarantee that the value is provided. In addition, it aligns with `hasBlockSupport` API methods from `@wordpress/blocks`. 

On the other hand, it might be more convenient to avoid it but then we need to ensure that the block editor loads all the defaults when there is no server.

gziolo replied:
We agreed to remove the `defaultValue` param for now.

oandregal reviewed (commented):
:wave: This is me thinking out loud about how all these pieces fit together, pardon me if you already know all of this, so I have a link for future reference. For clarity, I'll use `core.json` to refer to `lib/experimental-default-theme.json` and `theme.json` for the one provided by the theme.

So, we currently have:

- Block settings: persisted in the `core/blocks` store. There's a `blocks.registerBlockType` filter in the client to hook into this. This data comes from block.json.

- Editor settings: persisted in the `core/block-editor` store. There's a `block_editor_settings` filter in the server to hook into this. This data comes from PHP (several sources).

And we want:

1. Block settings:
    1. Default values for supports are taken from block.json.
    2. Via the `blocks.registerBlockType` client filter, we override the defaults with data taken form the `core/block-name` subtree (the result of merging core.json & theme.json).
    3. Via the `blocks.registerBlockType` client filter third-party code can still provide further overrides.

2. Editor settings:
    1. Default values for editor settings taken from PHP (several sources).
    2. Via the `block_editor_settings` server filter, we override the defaults with data taken from the `global` subtree (the result of merging core.json & theme.json).
    3. Via the `block_editor_settings` server filter, third-party code can still provide further overrides.

The `useEditorFeature` component merges the block & editor preferences in this way: if they exist, block settings take precedence over editor settings.

This PR implements 2.2, but we still need to implement 1.2 (it's fine if we do this in a follow-up PR).

nosolosw commented on lib/editor-features.php:
```diff
@@ -0,0 +1,49 @@
+ */
+function gutenberg_experimental_get_editor_features_config() {
+	$empty_config = array();
+	$config_path  = __DIR__ . '/experimental-default-theme.json';
```
As you suggested, I'm thinking that we can move the `theme.json` (both core and theme) to PHP. This also solves one potential code path in which we return the empty config without any values. We can do this in a follow-up as we also have to do it for global styles.

gziolo replied:
Let's take care of it separately. I'm still unsure whether this JSON file shouldn't be used as a default for `@wordpress/block-editor` settings object:

https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/store/defaults.js

There is also this question whether should I put some defaults there as well for features 😅

nosolosw commented on lib/editor-features.php:
```diff
@@ -0,0 +1,49 @@
+	}
+
+	$theme_config = json_decode(
+		@file_get_contents( $config_path ),
```
Is the `@` trick preferred to check if the file exists?

gziolo replied:
Related discussion:

https://github.com/WordPress/gutenberg/pull/22291#discussion_r427982610

The check for file existence is earlier, it prevents other potential issues with the file listed by @aduth  :)

nosolosw commented on packages/block-library/src/paragraph/block.json:
```diff
@@ -32,6 +32,11 @@
 		"__experimentalLineHeight": true,
-		"__experimentalFontSize": true
+		"__experimentalFontSize": true,
+		"__experimentalFeatures": {
```
I'd remove this `__experimentalFeatures` level, but I don't think this PR should be blocked by that. We can always iterate as more things are moved into this.

gziolo replied:
We can do it later once we settle on the final shape. For now, it makes it clear what is integrated with the block editor features.

gziolo commented:
> The `useEditorFeature` component merges the block & editor preferences in this way: if they exist, block settings take precedence over editor settings.

Correct.

> This PR implements 2.2, but we still need to implement 1.2 (it's fine if we do this in a follow-up PR).

At the moment we don't entirely implement 2.2, because `theme.json` (`experimental-theme.json`) isn't loaded, we load only features from `global` section of `core.json` (`lib/experimental-default-theme.json`).

In general, we need to decide separately how `theme.json` fits in this picture. For simplicity, it was removed from the scope in the final proposal.

oandregal reviewed (approved):
I'm approving this, so we can iterate on related things.

I'd still like if we could remove the extra default parameter added to `useEditorFeature`. It's not necessary for this feature just yet (if we remove it, this still works as expected). Once a few of the follow-up tasks are ready I think we won't need it, and having it now makes it more difficult to remove later.

These are the follow-up tasks that I see:

- [x] Editor Defaults: take & merge data from the `global` subtree of the `theme.json` (now it only takes core defaults).
- [x] Block Defaults: expand to this area. Ex: that a theme can disable the drop cap for the paragraph block from theme.json.
- [x] See how we can consolidate `lib/experimental-default-theme.json` and `block-editor/src/store/defaults.js`.
- [x] Document (can happen when a few more of the above things are ready).

gziolo committed:
Remove the default value param from useEditorFeature

gziolo commented:
Thank you for all feedback. I will open follow-up tasks shortly.

> Document (can happen when a few more of the above things are ready).

It's partially documented in https://github.com/WordPress/gutenberg/pull/22622 and we should continue adding new revisions as we build the whole experience.

gziolo commented:
Filed 3 follow up tasks as suggested in https://github.com/WordPress/gutenberg/pull/22291#pullrequestreview-418481941:
- #22655
- #22657
- #22658
